### ISSUE TITLE
Impose a custom format of SST

### ISSUE THREAD

**Issue by ShiKaiWi** [CeresDB/ceresdb#495]
### Describe This Problem

Currently, the format of sst is just parquet, which may lead to some possible issues in the future upgrading, e.g. introducing some new format like TSM to organize the data.

### Proposal

Impose the new custom format of SST, which may look like:
```
┌───────────────┐
│               │
│               │
│    Payload    │
│               │
│               │
├───────────────┤
│    Footer     │
└───────────────┘
```

And the footer should contain these fields:
```
/// The version number of the format for upgrading compatibility.
version: u16
/// Format of payload, e.g. Parquet, TSM
payload_format: u16
/// A number denotes that this is our custom format.
magical number: u32
```

So that the footer will be a fixed size 64B.

### Additional Context

Although this custom format brings us some great benefits, the sst can't be used by the third-party applications which can process the parquet files.

### Related work
Currently, the sst type is used to decide which type of sst reader/builder to create, but this option is configured for table level, that is to say, one table can only use one sst type. I guess this is not reasonable, and it is more flexible and elegant to allow the data of one table being encoded in different sst formats. So I guess we make this option only works when deciding the type of sst builder, and add a new sst type called `Auto` for the flexibility of choosing a better sst format.

**Comment by ShiKaiWi** [comment]
Here is the [TSM format](https://docs.influxdata.com/influxdb/v1.8/concepts/storage_engine/#tsm-files):
```
+--------+------------------------------------+-------------+--------------+
| Header |               Blocks               |    Index    |    Footer    |
|5 bytes |              N bytes               |   N bytes   |   4 bytes    |
+--------+------------------------------------+-------------+--------------+
+-------------------+
|      Header       |
+-------------------+
|  Magic  │ Version |
| 4 bytes │ 1 byte  |
+-------------------+
+---------+
│ Footer  │
+---------+
│Index Ofs│
│ 8 bytes │
+---------+
```

**Comment by ShiKaiWi** [comment]
Here is rocksdb's [BlockBasedTable Format](https://github.com/facebook/rocksdb/wiki/Rocksdb-BlockBasedTable-Format):
```
<beginning_of_file>
[data block 1]
[data block 2]
...
[data block N]
[meta block 1: filter block]                  (see section: "filter" Meta Block)
[meta block 2: index block]
[meta block 3: compression dictionary block]  (see section: "compression dictionary" Meta Block)
[meta block 4: range deletion block]          (see section: "range deletion" Meta Block)
[meta block 5: stats block]                   (see section: "properties" Meta Block)
...
[meta block K: future extended block]  (we may add more meta blocks in the future)
[metaindex block]
[Footer]                               (fixed size; starts at file_size - sizeof(Footer))
<end_of_file>
```

**Comment by ShiKaiWi** [comment]
Here is the [Parquet File Format](https://parquet.apache.org/docs/file-format/):
```
4-byte magic number "PAR1"
<Column 1 Chunk 1 + Column Metadata>
<Column 2 Chunk 1 + Column Metadata>
...
<Column N Chunk 1 + Column Metadata>
<Column 1 Chunk 2 + Column Metadata>
<Column 2 Chunk 2 + Column Metadata>
...
<Column N Chunk 2 + Column Metadata>
...
<Column 1 Chunk M + Column Metadata>
<Column 2 Chunk M + Column Metadata>
...
<Column N Chunk M + Column Metadata>
File Metadata
4-byte length in bytes of file metadata
4-byte magic number "PAR1"
```

**Comment by ShiKaiWi** [comment]
After discussions with @jiacai2050, we decide to store the file storage format in the manifest rather than imposing a custom storage format.

**Comment by ShiKaiWi** [comment]
Finally, self-describing SST will bring us many benefits compared with storing the file format in the manifest which leads to the dilemma that the separate SSTs can't be decoded.

**Comment by ShiKaiWi** [comment]
Fixed by #570
